test/controllers/tasks_controller_test.rb

```diff
@@ -105,24 +110,110 @@
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
     it "can update an existing task" do
-      skip
+      task = Task.create!(name: "original")
+      task_data = {
+        task: {
+          name: "changed",
+        },
+      }
+
+      # Act
+      patch task_path(task), params: task_data
+
+      # Assert
+      must_respond_with :redirect
+      must_redirect_to task_path(task)
+
+      task.reload
+      expect(task.name).must_equal(task_data[:task][:name])
+
       # Your code here
     end
 
     it "will redirect to the root page if given an invalid id" do
-      skip
+      task_data = {
+        task: {
+          name: "changed",
+        },
+      }
+      patch task_path(-1), params: task_data
+      must_respond_with :redirect
       # Your code here
     end
+
+    it "will not update if the params are invalid" do
+      id = task.id
+  
+      expect {
+        patch task_path(id), params: {}
+      }.must_raise ActionController::ParameterMissing
+
+    end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    
+    
+    it "can destroy a task" do
```

Comment:
You should also test what happens when you try to delete a task that doesn't exist.